### PR TITLE
KAFKA-5823: Extend upgrade section for KIP-120

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review \cc @guozhangwang @dguy @bbejeck 

(I did leak something about KIP-138, too, as it was easier this way IMHO. Hope that's ok.)

**Comment by Sharon2017** [comment]
A nit on line 54 in the file  docs/streams/upgrade-guide.html:
> old API in only deprecated
Nit: in -> is

File: docs/streams/upgrade-guide.html [hunk 1]
```diff
@@ -45,7 +45,44 @@
...
 
     <h3><a id="streams_api_changes_100" href="#streams_api_changes_100">Streams API changes in 1.0.0</a></h3>
-    <!-- TODO: KIP-120, 161, 167, 173 -->
+    <!-- TODO: 161, 167, 173, 182 -->
+
+    <p>
+        With 1.0 a major API refactoring was accomplished and the new API is cleaner and easier to use.
+        This change includes the five main classes <code>KafakStreams<code>, <code>KStreamBuilder</code>,
+        <code>KStream</code>, <code>KTable</code>, and <code>TopologyBuilder</code> (and some more others).
+        All changes are fully backward compatible as old API in only deprecated but not removed.
```
Comment (dguy) [inline docs/streams/upgrade-guide.html]: old API is deprecated

File: docs/streams/upgrade-guide.html [hunk 2]
```diff
@@ -45,7 +45,44 @@
...
+        or the Processor API (<code>TopologyBuilder</code>) were deprecated and replaced by
+        <code>StreamsBuilder</code> and <code>Topology<code> (both new classes are located in
+        package <code>org.apache.kafka.streams</code>.
+        Note, that <code>StreamsBuilder</code> does not extend <code>Topology</code>, i.e.,
+        the class hierarchy is different now.
+        The new classes have basically the same methods as the old ones to build a topology via DSL or Processor API.
+        However, some internal methods that were public in <code>KStreamBuilder</code>
+        and <code>TopologyBuilder</code> but not part of the actual API are not present
+        in the new classe any longer.
+        Furthermore, some overloads got simplified compared to the original classes.
```
Comment (dguy) [inline docs/streams/upgrade-guide.html]: got -> were

**Review by dguy — COMMENTED** [review]
Thanks @mjsax, just a couple of comments

**Comment by mjsax** [comment]
Thanks @Sharon2017 @dguy 

Updated this.

File: docs/streams/upgrade-guide.html [hunk 3]
```diff
@@ -45,7 +45,44 @@
...
+        All changes are fully backward compatible as old API is only deprecated but not removed.
+        We recommend to move to the new API as soon as you can.
+        We will summarize all API changes in the next paragraphs.
+    </p>
+
+    <p>
+        The two main classes to specify a topology via the DSL (<code>KStreamBuilder</code>)
+        or the Processor API (<code>TopologyBuilder</code>) were deprecated and replaced by
+        <code>StreamsBuilder</code> and <code>Topology<code> (both new classes are located in
+        package <code>org.apache.kafka.streams</code>.
```
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: `)` missing?

File: docs/streams/upgrade-guide.html [hunk 4]
```diff
@@ -45,7 +45,44 @@
...
+
+    <p>
+        Changing how a topology is specified also affects <code>KafkaStreams<code> constructors,
+        that now only accept a <code>Topology</code>.
+        Using the DSL builder class <code>StreamsBuilder</code> one can get the constructed
+        <code>Topology</code> via <code>StreamsBuilder#build()</code>.
+        Additionally, a new class <code>org.apache.kafka.streams.TopologyDescription</code>
+        (and some more dependent classes) were added.
+        Those can be used to get a detailed description of the specified topology
+        and can be obtained by calling <code>Topology#describe()</code>.
```
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: Could reference to the quickstart section as an example.

**Review by guozhangwang — COMMENTED** [review]
Minor comments, otherwise LGTM.

**Comment by mjsax** [comment]
Updated.

**Comment by guozhangwang** [comment]
Merged to trunk, thanks @mjsax !

File: docs/streams/upgrade-guide.html [hunk 5]
```diff
@@ -45,7 +45,45 @@
     </p>
 
     <h3><a id="streams_api_changes_100" href="#streams_api_changes_100">Streams API changes in 1.0.0</a></h3>
-    <!-- TODO: KIP-120, 161, 167, 173 -->
+    <!-- TODO: 161, 167, 173, 182 -->
+
+    <p>
+        With 1.0 a major API refactoring was accomplished and the new API is cleaner and easier to use.
+        This change includes the five main classes <code>KafakStreams<code>, <code>KStreamBuilder</code>,
```
Comment (kamalcph) [inline docs/streams/upgrade-guide.html:52]: @mjsax 
nit: KafakStreams

File: docs/streams/upgrade-guide.html [hunk 6]
```diff
@@ -45,7 +45,45 @@
...
+        The two main classes to specify a topology via the DSL (<code>KStreamBuilder</code>)
+        or the Processor API (<code>TopologyBuilder</code>) were deprecated and replaced by
+        <code>StreamsBuilder</code> and <code>Topology<code> (both new classes are located in
+        package <code>org.apache.kafka.streams</code>).
+        Note, that <code>StreamsBuilder</code> does not extend <code>Topology</code>, i.e.,
+        the class hierarchy is different now.
+        The new classes have basically the same methods as the old ones to build a topology via DSL or Processor API.
+        However, some internal methods that were public in <code>KStreamBuilder</code>
+        and <code>TopologyBuilder</code> but not part of the actual API are not present
+        in the new classe any longer.
```
Comment (kamalcph) [inline docs/streams/upgrade-guide.html:69]: @mjsax 
nit: classe

**Comment by mjsax** [comment]
Thanks @Kamal15 ! I did open https://github.com/apache/kafka/pull/3804

**Pull request closed without merging**